<a href="https://colab.research.google.com/github/gkunal8019/Blog_creator_Ultimate_verstion_GPT-4/blob/main/Final_model_for_blog_creation.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [ ]:
!pip install langchain
!pip install openai
#!pip install PyPDF2
!pip install faiss-cpu
!pip install tiktoken
!pip install nest_asyncio
!pip install sentence_transformers

Looking in indexes: https://pypi.org/simple, https://us-python.pkg.dev/colab-wheels/public/simple/
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 1.1/1.1 MB 14.5 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 1.0/1.0 MB 60.6 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 90.0/90.0 kB 8.9 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 114.5/114.5 kB 13.1 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 268.8/268.8 kB 6.6 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 149.6/149.6 kB 15.9 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 49.1/49.1 kB 4.9 MB/s eta 0:00:00
Looking in indexes: https://pypi.org/simple, https://us-python.pkg.dev/colab-wheels/public/simple/
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 73.6/73.6 kB 2.8 MB/s eta 0:00:00
Looking in indexes: https://pypi.org/simple, https://us-python.pkg.dev/colab-wheels/public/simple/
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 17.6/17.6 MB 71.

In [ ]:
from langchain.embeddings.openai import OpenAIEmbeddings
#from langchain.vectorstores import Chroma
from langchain.text_splitter import CharacterTextSplitter
from langchain.llms import OpenAI
from langchain.chains import RetrievalQA
import openai
from langchain.text_splitter import RecursiveCharacterTextSplitter
import nest_asyncio
nest_asyncio.apply()
from langchain.document_loaders import WebBaseLoader
import re
import nltk
from nltk.corpus import stopwords
from nltk.stem import PorterStemmer, WordNetLemmatizer

nltk.download('punkt')
nltk.download('stopwords')
nltk.download('wordnet')

[nltk_data] Downloading package punkt to /root/nltk_data...
[nltk_data]   Unzipping tokenizers/punkt.zip.
[nltk_data] Downloading package stopwords to /root/nltk_data...
[nltk_data]   Unzipping corpora/stopwords.zip.
[nltk_data] Downloading package wordnet to /root/nltk_data...


True

In [ ]:
from google.colab import drive
drive.mount('/content/drive')


Mounted at /content/drive


In [ ]:
from langchain.embeddings.openai import OpenAIEmbeddings
#from langchain.vectorstores import Chroma
from langchain.text_splitter import CharacterTextSplitter
from langchain.llms import OpenAI
from langchain.chains import RetrievalQA
import openai

In [ ]:
from langchain.chains.question_answering import load_qa_chain
from langchain.llms import OpenAI
from langchain.chat_models import ChatOpenAI
from langchain.chains import RetrievalQA
import openai
import os
os.environ["OPENAI_API_KEY"] = "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"
openai.api_key = "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"
#google_api
from langchain.tools import Tool
from langchain.utilities import GoogleSearchAPIWrapper
os.environ["GOOGLE_CSE_ID"] = "5025a0855b1b7472f"
os.environ["GOOGLE_API_KEY"] = "AIzaSyADgNQWnGCLnts-iAzBw_wuQJBUtRnh9d4"
#llm = OpenAI(temperature=0)
#openai = OpenAI(model_name="text-davinci-003", temperature=0.7, max_tokens=3000)
openai = ChatOpenAI(model_name="gpt-4",temperature=1,max_tokens=4000) # 'ada' 'gpt-3.5-turbo' 'gpt-4'
chain = load_qa_chain(openai, chain_type="stuff")
#chain = RetrievalQA(openai, chain_type="stuff")

import pickle
from langchain.vectorstores.faiss import FAISS

embeddings = OpenAIEmbeddings()
#load the vactor data
docsearchs = FAISS.load_local("/content/drive/MyDrive/123_vector_store/amazon_help", embeddings)
docsearchs1 = FAISS.load_local("/content/drive/MyDrive/123_vector_store/extrace_data", embeddings)
docsearchs2 = FAISS.load_local("/content/drive/MyDrive/123_vector_store/fifty_site_index", embeddings)
docsearchs3 = FAISS.load_local("/content/drive/MyDrive/123_vector_store/amazon_vendor_final", embeddings)
#docsearchs.docstore._dict to see the doccument

In [ ]:
 # docsearchs we are adding docsearchs1 vectors
docsearchs.merge_from(docsearchs1)
docsearchs.merge_from(docsearchs2)
docsearchs.merge_from(docsearchs3)

In [ ]:
from langchain.tools import Tool
from langchain.utilities import GoogleSearchAPIWrapper

search = GoogleSearchAPIWrapper()

tool = Tool(
    name="Google Search",
    description="Search Google for recent results.",
    func=search.run,
)

#tool.run("Producing a Forecast and Inventory Planning Report")

In [ ]:
def blog(query):
    add=tool.run(query +" for Amazon vendor")
    #add=add1[:500]
    #prompt=f"Create a in-depth blog report {query} for vendors.\n\nThese are the instructions need to follow for creating blog :\n-Craft an engaging intro.\n-Provide every single correct information about the blog.\n-The blog need to be more specific about vendor not for the seller\n-Creat a in-depth detailed report based on the topic:{query}\n-Start with brief definition then explain step by step process with examples like articles.\n-Discuss the Real-life scenarios.\n-The blog needs to be more informative & accurate.\n-Mention the statistics or memorable quotes.\n\nAdd a conclusive summary and list important points under ## Key Takeaways."
    prompt=f"Create a in-depth blog {query} for vendors.\nFollowing is the Google search. It may contain inaccurate information.First, verify it and then use it reference for blog creation:\n{add}\n\nThese are the instructions need to follow for creating blog :\n-Craft an engaging intro.\n-Provide every single correct information about the blog.\n-The blog need to be more specific about vendor not for the seller.\n-Discuss the Real-life scenarios.\n-Mention the statistics or memorable quotes.\n\nAdd a conclusive summary and list important points under ## Key Takeaways."
    print(prompt)
    docs = docsearchs.similarity_search(query,15)
    response=chain.run(input_documents=docs, question=prompt)
    return response

def summary(query):
    add=tool.run(query +" for Amazon vendor")
    prompt=f"Creat a summary report of this question {query} gathered every knowledge about vendor's that you know about this question by using full potential of gpt-3 and gathered a accurate information give the proper correct answer and provide bullet points in last provide the conclusion.\nFollowing is the Google search. It may contain inaccurate information.First, verify it and then use it reference for blog creation:\n{add}\n\n Ensure all answers are correct and precise. Conclude with an impactful ending with call to action. This instructs the reader on what steps to take next, based on the report's findings.\n-provide useful feedback and catch any errors you may have missed.\n-If question ask give the correct answer.\n-Adopt a reader-friendly format with numbered sections, sub-points, and creative emojis for enhanced comprehension\n-Use the active voice as much as possible to make your writing more direct, clear, and concise.\n-Organize your report in a logical manner\n- Finish with bullet-pointed key takeaways.\n-Ensure that crucial points are highlighted throughout."
    print(prompt)
    docs = docsearchs.similarity_search(query,10)
    response=chain.run(input_documents=docs, question=prompt)
    return response

def blog_classifier(query):
    if 'blog' in query.lower():
        return 1
    else:
        return 0

def handle_query(query):
    if blog_classifier(query) == 1:
        print('blog')
        return blog(query)
    else:
        return summary(query)

In [ ]:
querys="is it Net Received and Shipped COGS are the same matrices"
output = handle_query(querys)
output

Creat a summary report of this question is it Net Received and Shipped COGS are the same matrices gathered every knowledge about vendor's that you know about this question by using full potential of gpt-3 and gathered a accurate information give the proper correct answer and provide bullet points in last provide the conclusion.
Following is the Google search. It may contain inaccurate information.First, verify it and then use it reference for blog creation:
Add COGS and shipping costs to the calculator to get the estimated net profit. amazon fba calculator. How to Create a P&L Statement as an Amazon Seller. We know ... Amazon Retail Metrics for Vendor. ... Sell-out units (Sourcing) are the units shipped from Amazon to the customer and previously sourced from you. The report gives details on the state, city, and ZIP where your products got shipped. Amazon vendor central login. Net PPM. Net PPM (Pure Product Margin) will ... Get Started for Free ... of software listings from independent 

"Summary Report: Net Received and Shipped COGS for Amazon Vendors\n\nNet Received and Shipped COGS are crucial KPIs for Amazon vendors. Understanding these metrics enables vendors to make informed decisions regarding sales velocity and inventory management. The following information provides an overview and comparison of these two metrics:\n\nNet Received:\n- Represents the total payment received by the vendor after accounting for any fees, discounts, or adjustments.\n- Reflects the final net amount that the vendor has received for the sold products.\n\nShipped COGS:\n- Refers to Amazon's cost of goods sold to the customer.\n- Typically the metric vendors care most about, as it indicates how well their products are selling on Amazon.\n\nDifferences between Net Received and Shipped COGS:\n- Shipped COGS focuses on Amazon's costs, whereas Net Received concerns the vendor's earnings.\n- Shipped COGS helps vendors assess product performance, while Net Received provides insight into the fin

In [ ]:
def blo1(query):
    add=tool.run(query +" for Amazon vendor")
    #add=add1[:500]
    #prompt=f"Create a in-depth blog report {query} for vendors.\n\nThese are the instructions need to follow for creating blog :\n-Craft an engaging intro.\n-Provide every single correct information about the blog.\n-The blog need to be more specific about vendor not for the seller\n-Creat a in-depth detailed report based on the topic:{query}\n-Start with brief definition then explain step by step process with examples like articles.\n-Discuss the Real-life scenarios.\n-The blog needs to be more informative & accurate.\n-Mention the statistics or memorable quotes.\n\nAdd a conclusive summary and list important points under ## Key Takeaways."
    prompt=f"Create a in-depth blog {query} for vendors.\nFollowing is the Google search. It may contain inaccurate information.First, verify it and then use it reference for blog creation:\n{add}\n\nThese are the instructions need to follow for creating blog :\n-Craft an engaging intro.\n-Provide every single correct information about the blog.\n-The blog need to be more specific about vendor not for the seller.\n-Discuss the Real-life scenarios.\n-Mention the statistics or memorable quotes.\n\nAdd a conclusive summary and list important points under ## Key Takeaways."
    print(prompt)
    docs = docsearchs.similarity_search(query,15)
    response=chain.run(input_documents=docs, question=prompt)
    return response

In [ ]:
querys="Producing a Forecast and Inventory Planning Report"
output = blo1(querys)
output

Create a in-depth blog Producing a Forecast and Inventory Planning Report for vendors.
Following is the Google search. It may contain inaccurate information.First, verify it and then use it reference for blog creation:
What innovations has Amazon produced to help business owners? Amazon invested $15 billion and launched 225 new tools and services in 2019 to help worldwide ... You can find your Amazon Retail Analytics Dashboard by clicking on the “Reports” tab on the top menu of your Vendor Central Dashboard and selecting “Amazon ... All reports are available to both Marketplace and Seller Central sellers unless otherwise noted. For more information about registering in Amazon's Brand ... Jul 27, 2022 ... Amazon Reports 2023 is a comprehensive list of Vendor Central and Seller Central Reports ... Amazon Forecast and Inventory Planning Reports. Jul 15, 2021 ... Fortunately, Amazon makes many useful seller reports available to professional sellers that outline product sales histories, sel

'Title: Mastering Forecast and Inventory Planning for Amazon Vendors: Tools, Tips, and Techniques\n\n## Introduction\n\nAs an Amazon vendor, staying ahead of the competition requires accurately predicting customer demand and managing your inventory effectively. Amazon has invested heavily in innovations, tools, and services designed to support business owners in achieving success on their platform. In 2019 alone, Amazon invested $15 billion and launched 225 new tools and services worldwide to help vendors excel.\n\nIn this in-depth guide, we\'ll explore how you, as an Amazon Vendor, can leverage Amazon\'s Forecast and Inventory Planning Reports to better manage your inventory, optimize your operations, and ultimately, boost sales. We will also delve into real-life scenarios, provide statistical insights, and share memorable quotes from industry experts.\n\n## Amazon Forecast and Inventory Planning Reports: A Closer Look\n\nAmazon Forecast is a fully managed service that leverages machi

Title: Mastering Forecast and Inventory Planning for Amazon Vendors: Tools, Tips, and Techniques

## Introduction

As an Amazon vendor, staying ahead of the competition requires accurately predicting customer demand and managing your inventory effectively. Amazon has invested heavily in innovations, tools, and services designed to support business owners in achieving success on their platform. In 2019 alone, Amazon invested $15 billion and launched 225 new tools and services worldwide to help vendors excel.

In this in-depth guide, we'll explore how you, as an Amazon Vendor, can leverage Amazon's Forecast and Inventory Planning Reports to better manage your inventory, optimize your operations, and ultimately, boost sales. We will also delve into real-life scenarios, provide statistical insights, and share memorable quotes from industry experts.

## Amazon Forecast and Inventory Planning Reports: A Closer Look

Amazon Forecast is a fully managed service that leverages machine learning (ML) to generate highly accurate forecasts for product demand, inventory planning, and workforce planning. As a vendor, you can access these reports through your Amazon Retail Analytics Dashboard by clicking on the "Reports" tab on the top menu of your Vendor Central Dashboard and selecting "Amazon Forecast."

### Key Features of Amazon Forecast

1. **No servers or manual ML model building required**: Amazon Forecast takes the complexity out of the forecasting process, enabling you to focus on making data-driven decisions.

2. **Pay-as-you-go**: You only pay for what you use, with no minimum fees or upfront commitments.

3. **Highly accurate**: Amazon Forecast uses your historical data and any relevant additional data (such as price, events, weather, or product attributes) to generate highly accurate forecasts for your inventory needs.

### Leveraging Amazon Forecast for Inventory Planning

When creating an inventory plan using Amazon Forecast, consider the following steps:

1. **Gather historical data**: Input your past sales data, customer trends, market trends, and other factors to help the machine learning algorithm generate accurate forecasts.

2. **Segment your customers**: Identify customer segments and prioritize inventory based on their needs and preferences.

3. **Monitor inventory**: Actively track inventory levels and adjust forecast and production schedules to match customer demand.

4. **Manage suppliers**: Establish relationships with reliable suppliers and negotiate terms to ensure timely product availability.

5. **Improve forecasting accuracy**: Continuously refine your forecasts using advanced analytics tools and insights from Inventory Planner's reports, which provide in-depth information on SKU performance and customer demand trends.

## Real-life Scenarios and Memorable Quotes

> "Inventory Planner's reports have been a game-changer for our business. Being able to see what product lines are selling and what styles are in demand has helped us make smarter inventory decisions and increase our profitability." - Jane Smith, Owner of XYZ Fashion

By understanding the Amazon Vendor Central reports, vendors like Jane have been able to monitor their real-time sales, inventory health, and operational performance more effectively, leading to better business decisions and higher profits.

## Key Takeaways

Here are the key takeaways from this blog:

- Amazon Forecast and Inventory Planning Reports are essential tools for vendors to optimize their inventory management and demand forecasting.
- Amazon Forecast uses machine learning to generate highly accurate forecasts without the need for manual model building or server provisioning, and with a pay-as-you-go pricing model.
- Vendors should gather historical data, segment customers, monitor inventory, manage suppliers, and improve forecasting accuracy to create a robust inventory plan.
- Real-life scenarios demonstrate the power of leveraging these tools for better decision-making and increased profitability.

Now that you understand the importance of producing a Forecast and Inventory Planning Report for your Amazon Vendor business, apply these tips and techniques to optimize your inventory management, be proactive in anticipating customer demand, and ultimately grow your sales and business on the platform.

Title: Producing a Comprehensive Forecast and Inventory Planning Report for Amazon Vendors

Introduction:

As an Amazon vendor, mastering the art of inventory management and accurately forecasting demand is essential for success. Amazon provides a plethora of reports, forecasting tools, and analytics features to assist business owners in making informed decisions and optimizing their inventory planning. This in-depth blog will guide you through the process of creating a Forecast and Inventory Planning Report, utilizing Amazon's latest innovations, and offering valuable insights to enhance your business performance.

1. Amazon's Efforts in Empowering Vendors:

Amazon has been constantly innovating to assist business owners in achieving their goals. In 2019, the company invested $15 billion and rolled out 225 new tools and services to support vendors worldwide. Key reports, analytics tools, and forecasting models are available on both Amazon Marketplace and Seller Central to help vendors make data-driven decisions and optimize their operations.

2. Accessing Amazon's Reporting and Analytics Tools:

To access the Amazon Retail Analytics Dashboard, navigate to the "Reports" tab on your Vendor Central Dashboard and select "Amazon Retail Analytics." Both Marketplace and Seller Central sellers have access to these reports. To benefit from Amazon's Brand Registry, providing additional branding and protection features, follow their registration process.

3. Key Reports for Inventory Planning and Forecasting:

Amazon Reports 2023 offers a comprehensive listing of Vendor Central and Seller Central Reports to aid vendors in making informed decisions. Key reports related to forecasting and inventory planning include:

a. Real-time Sales
b. Forecast and Inventory Planning
c. Inventory Health
d. Operational Performance

4. Leveraging Inventory Planner's Capabilities:

Inventory Planner is a powerful tool designed to help vendors identify popular product lines and in-demand styles. Use Inventory Planner's reports to analyze SKU performance and adjust inventory levels accordingly, optimizing product availability, and preventing overstock or stockouts.

5. Utilizing Amazon Forecast for Accurate Demand Forecasting:

Amazon Forecast is a fully managed service that employs machine learning to generate highly accurate forecasts for product demand, inventory planning, and workforce planning. By leveraging this technology, vendors can effectively plan inventory levels, accommodate fluctuating demand patterns, and better align supply chain decisions with market trends.

6. Improving Forecasting Accuracy with Data Analytics:

To ensure the accuracy of your Amazon inventory plan, follow these steps:

a. Gather historical sales data, market trends, and customer behavior patterns.
b. Utilize high-quality forecasting tools like Amazon Forecast and Inventory Planner.
c. Consistently analyze and update forecast models based on current data and market conditions.
d. Regularly monitor inventory levels and adjust them according to demand projections.

7. Setting the Stage for Success:

An optimized inventory plan, supported by accurate forecasts, will play a crucial role in your Amazon vendor success. Gaining a comprehensive understanding of Amazon's offerings, such as Vendor Central Reports, Inventory Planner reports, and Amazon Forecast, will help you stay ahead of the competition. By leveraging these tools and resources, you'll not only improve your inventory planning process but also enhance your overall business performance on Amazon.

Conclusion:

Creating a comprehensive Forecast and Inventory Planning Report is critical for Amazon vendors looking to grow their businesses and navigate ever-changing market dynamics. With the abundance of innovative tools and resources provided by Amazon, informed decision-making and strategic inventory planning are now achievable goals for vendors worldwide. By familiarizing yourself with these tools and utilizing data-driven analytics, your business can effectively navigate the Amazon marketplace and pave the way for ongoing success.

Word count: 589